library/std/src/sys/windows/path.rs

```diff
@@ -50,21 +50,79 @@ pub(crate) fn append_suffix(path: PathBuf, suffix: &OsStr) -> PathBuf {
     path.into()
 }
 
+struct PrefixParser<'a> {
+    path: &'a OsStr,
+    prefix: Vec<u8>,
+}
+
+impl<'a> PrefixParser<'a> {
+    fn get_prefix(path: &OsStr, prefix_len: usize) -> Vec<u8> {
+        let mut prefix = path.bytes()[..prefix_len.min(path.len())].to_owned();
+        // SAFETY: Only ASCII characters are modified.
+        for ch in &mut prefix {
+            if ch == &b'/' {
+                *ch = b'\\';
+            }
+        }
+        prefix
+    }
+
+    fn new(path: &'a OsStr, prefix_len: usize) -> Self {
+        Self { path, prefix: Self::get_prefix(path, prefix_len) }
+    }
+
+    fn as_slice(&self) -> PrefixParserSlice<'a, '_> {
+        PrefixParserSlice { parser: self, index: 0 }
+    }
+}
+
+struct PrefixParserSlice<'a, 'b> {
+    parser: &'b PrefixParser<'a>,
+    index: usize,
+}
+
+impl<'a> PrefixParserSlice<'a, '_> {
+    fn strip_prefix(&self, prefix: &str) -> Option<Self> {
+        self.parser.prefix[self.index..]
+            .starts_with(prefix.as_bytes())
+            .then(|| Self { parser: self.parser, index: self.index + prefix.len() })
+    }
+
+    fn prefix_bytes(&self) -> &'a [u8] {
+        &self.parser.path.bytes()[..self.index]
+    }
+
+    fn finish(self) -> &'a OsStr {
+        // SAFETY: The unsafety here stems from converting between &OsStr and
+        // &[u8] and back. This is safe to do because (1) we only look at ASCII
+        // contents of the encoding and (2) new &OsStr values are produced only
+        // from ASCII-bounded slices of existing &OsStr values.
+        unsafe { bytes_as_os_str(&self.parser.path.bytes()[self.index..]) }
+    }
+}
+
 pub fn parse_prefix(path: &OsStr) -> Option<Prefix<'_>> {
     use Prefix::{DeviceNS, Disk, Verbatim, VerbatimDisk, VerbatimUNC, UNC};
 
-    if let Some(path) = strip_prefix(path, r"\\") {
+    let parser = PrefixParser::new(path, 8);
```

Comment:
Since prefix is relatively small, I wonder if storing it in a `[u8; 8]` instead of a `Vec<u8>` might be better? It would save an allocation

Reply:
Right, I should use const generics here.